### PR TITLE
WIP: GModel dump POC in G-API CPU backend.

### PR DESCRIPTION
**Pull request by dbudniko**
<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
buildworker:Custom=linux-1
build_image:Custom=ubuntu-clang:18.04
buildworker:Mac=macosx-2
```

### CODE REVIEW

File: modules/gapi/include/opencv2/gapi/gtype_traits.hpp [hunk 1]
```diff
@@ -41,6 +41,30 @@ namespace detail
...
+    template<typename T> struct GOpaqueTraits;
+    template<typename T> struct GOpaqueTraits { static constexpr const OpaqueKind kind = OpaqueKind::UNSUPPORTED; };
+    template<> struct GOpaqueTraits<int> { static constexpr const OpaqueKind kind = OpaqueKind::INT; };
+    template<> struct GOpaqueTraits<double> { static constexpr const OpaqueKind kind = OpaqueKind::DOUBLE; };
+    template<> struct GOpaqueTraits<cv::Size> { static constexpr const OpaqueKind kind = OpaqueKind::CV_SIZE; };
+    template<> struct GOpaqueTraits<bool> { static constexpr const OpaqueKind kind = OpaqueKind::BOOL; };
+    template<> struct GOpaqueTraits<cv::Scalar> { static constexpr const OpaqueKind kind = OpaqueKind::CV_SCALAR; };
+    template<> struct GOpaqueTraits<cv::Point> { static constexpr const OpaqueKind kind = OpaqueKind::CV_POINT; };
+    template<> struct GOpaqueTraits<cv::Mat> { static constexpr const OpaqueKind kind = OpaqueKind::CV_MAT; };
+    template<> struct GOpaqueTraits<cv::Rect> { static constexpr const OpaqueKind kind = OpaqueKind::CV_RECT; };
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/gtype_traits.hpp]: Is this class relevant to `cv::GOpaque<T>` introduced by @smirnov-alexey ?

File: modules/gapi/src/backends/cpu/gcpubackend.cpp [hunk 2]
```diff
@@ -73,9 +103,19 @@ cv::gapi::GBackend cv::gapi::cpu::backend()
...
 {
+    //const auto s = serialization::serialize(m_gm, nodes);
+    //serialization::deserialize(s);
+    //serialization::printGSerialized(s);
+
+    // FIXME: reuse code from GModelBuilder/GModel!
+    // ObjectCounter?? (But seems we need existing mapping by shape+id)
+
+
+
```
Comment (dmatveev) [inline modules/gapi/src/backends/cpu/gcpubackend.cpp]: Not sure if this code is required here

File: modules/gapi/src/backends/cpu/serialization.cpp [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+
+    for (const auto &out_nh : nh->outNodes())
+    {
+        putData(s, cg, out_nh);
+    }
+
+    s.m_ops.push_back(sop);
+}
+
+void printOp(const Op& op)
```
Comment (dmatveev) [inline modules/gapi/src/backends/cpu/serialization.cpp]: please add the output stream as an `ostream&` parameter and don't refer to `std::cout` directly

File: modules/gapi/src/backends/cpu/serialization.cpp [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+    auto nh = g.createNode();
+    GModel::Graph gm(g);
+
+    gm.metadata(nh).set(NodeType{NodeType::DATA});
+
+    HostCtor ctor{};
+    // internal?
+    auto storage = gimpl::Data::Storage::INTERNAL;
+    gm.metadata(nh).set(gimpl::Data{data.rc.shape, data.rc.id, data.meta, ctor, storage});
+}
```
Comment (dmatveev) [inline modules/gapi/src/backends/cpu/serialization.cpp]: Does this differ much from what's defined in `GModel`?

File: modules/gapi/src/backends/cpu/serialization.hpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+struct Data
+{
+    // GModel::Data consists of shape+(int)rc
+    RcDesc rc;
+    GMetaArg meta;
+    // storage?
+
+    bool operator==(const Data& d) const { return rc == d.rc; }
+};
```
Comment (dmatveev) [inline modules/gapi/src/backends/cpu/serialization.hpp]: The above structures also duplicate what's defined in GModel?

**Comment by alalek** [comment]
@dbudniko Please disable excessive logging in tests.

File: modules/gapi/test/common/gapi_compoundkernel_tests.cpp [hunk 6]
```diff
@@ -443,7 +443,7 @@ TEST(GCompoundKernel, CompoundExpandedToCompound)
 
     EXPECT_EQ(0, cv::countNonZero(out_mat != ref_mat));
 }
-
+#if 0
```
Comment (smirnov-alexey) [inline modules/gapi/test/common/gapi_compoundkernel_tests.cpp]: What's wrong with the GArray?

Comment (smirnov-alexey) [inline modules/gapi/test/common/gapi_compoundkernel_tests.cpp]: Why so many tests are commented?

File: modules/gapi/src/backends/cpu/serialization.cpp [hunk 7]
```diff
@@ -0,0 +1,876 @@
...
+        std::cout << op.k.tag << std::endl;
+    }
+
+    std::cout << "printOp  Args" << std::endl;
+    size_t i_int = 0; size_t i_double = 0; size_t i_size = 0; size_t i_bool = 0;
+    size_t i_scalar = 0; size_t i_point = 0; size_t i_mat = 0; size_t i_rect = 0;
+    size_t i_objref = 0; size_t i_opaque = 0;
+    for (const auto& kind : op.kind)
+    {
+        if(kind == (int)detail::ArgKind::GOBJREF)
```
Comment (smirnov-alexey) [inline modules/gapi/src/backends/cpu/serialization.cpp]: Better to use `static_cast` here and everywhere else

File: modules/gapi/src/backends/cpu/serialization.cpp [hunk 8]
```diff
@@ -0,0 +1,876 @@
...
+    uint ops_size;
+    ifs_ops.read((char*)&ops_size, sizeof(uint));
+//    std::cout << "readGSerializedOps ops size " << ops_size <<  std::endl;
+    for (uint i = 0; i < ops_size;  i++)
+    {
+        //Kernel k
+        uint k_name_size;
+        ifs_ops.read((char*)&k_name_size, sizeof(uint));
+//        std::cout << "readGSerializedOps k_name_size " << k_name_size <<  std::endl;
+        std::string k_name;
```
Comment (smirnov-alexey) [inline modules/gapi/src/backends/cpu/serialization.cpp]: You can use constructor `k_name(k_name_size)` and remove `resize` (also applicable in several other places)

**Comment by asmorkalov** [comment]
@dbudniko @rgarnov @dmatveev  @smirnov-alexey Friendly reminder.

**Comment by asmorkalov** [comment]
@dmatveev @rgarnov @smirnov-alexey Any comments on the last commit?

**Comment by asmorkalov** [comment]
@dmatveev @rgarnov @smirnov-alexey Friendly reminder.

**Comment by dmatveev** [comment]
@dbudniko I believe we have another branch for this now, right? Can this PR be closed?

**Comment by dmatveev** [comment]
@dbudniko I believe this MR can be closed

**Pull request closed without merging**